[MRG] speed up `SeqToHashes` `translate`

mr-eyes opened the pull request:
Vectors are very good in insertions, not in deletion. In #1938 I replaced `Vec` with `VecDeque` because I was popping out a hash per iteration. This deletion is very slow in `Vec` because it shifts all the elements in memory, while it's very fast in `VecDeque`. 

In this PR I am reverting back to `Vec` and replacing all the parts that require deletion or popping by smoother code. Now there's no need for using `VecDeque`.

Resolves #1945

mr-eyes committed:
lightning fast translate :zap:

mr-eyes committed:
rust clippy

mr-eyes commented:
Ready for review @ctb @luizirber 
I believe it's better now :)

before and after timing in #1945

ctb reviewed (approved):
This is great! But I have to admit I don't understand why the new code is so much faster 😭 . Could you maybe add some comments?

mr-eyes commented:
@ctb sure!

Here I am explaining the tradeoff of using `Vec` vs `VecDeque` and why I reverted back to `Vec`.

> Vectors are very good in insertions, not in deletion. In #1938 I replaced `Vec` with `VecDeque` because I was popping out a hash per iteration. This deletion is very slow in `Vec` because it shifts all the elements in memory, while it's very fast in `VecDeque`.
> 
> In this PR I am reverting back to `Vec` and replacing all the parts that require deletion or popping by smoother code. Now there's no need for using `VecDeque`.

---

`SeqToHashes` is mainly an iterator (like a Python generator); In every iteration it hashes and return a kmer. Only in the `translate` mode, I hash all the kmers for all frames at once, and then consume a hash per iteration. The consuming/yelding part was the problem, I was popping an item in every yield until the vector of hashes is empty (this is very slow in vectors). In this PR I replaced the popping out with an vector indeces iterator and it will keep moving through the vector until it reaches the last element. Hope I well explained it :)

https://github.com/sourmash-bio/sourmash/blob/6f5245bd3357ca33207c00384bd9d116fe3ce5de/src/core/src/signature.rs#L226

https://github.com/sourmash-bio/sourmash/blob/6f5245bd3357ca33207c00384bd9d116fe3ce5de/src/core/src/signature.rs#L341-L348

ctb commented:
ahh, got it! but - is this something that is worth adding in comments in the actual code, do you think?

ctb commented:
(I tend to think that something that required the amount of discussion this issue had is worth documenting in the code.)

mr-eyes commented:
I spent sometime to understand what I was doing in the code, so, yes I will add some comments.

mr-eyes committed:
in-line comments for translate

mr-eyes commented:
@ctb I think it's ready now.

ctb commented:
cool! I mean, it's been approved for a while, but I can do the button push ;)

mr-eyes commented:
@ctb oh, ok :D next time I will merge when approved :) Thank you!
